In [1]:
import pandas as pd
import numpy as np

In [2]:
sales = pd.read_csv(r'C:\Users\use\Desktop\data viz\Diwali Sales Data.csv', encoding='unicode_escape')
sales

User_ID    Cust_name Product_ID Gender Age Group  Age  Marital_Status  \
0      1002903    Sanskriti  P00125942      F     26-35   28               0   
1      1000732       Kartik  P00110942      F     26-35   35               1   
2      1001990        Bindu  P00118542      F     26-35   35               1   
3      1001425       Sudevi  P00237842      M      0-17   16               0   
4      1000588         Joni  P00057942      M     26-35   28               1   
...        ...          ...        ...    ...       ...  ...             ...   
11246  1000695      Manning  P00296942      M     18-25   19               1   
11247  1004089  Reichenbach  P00171342      M     26-35   33               0   
11248  1001209        Oshin  P00201342      F     36-45   40               0   
11249  1004023       Noonan  P00059442      M     36-45   37               0   
11250  1002744      Brumley  P00281742      F     18-25   19               0   

                State      Zone       Occupation Product_Category  Orders  \
0         Maharashtra   Western       Healthcare             Auto       1   
1      Andhra Pradesh  Southern             Govt             Auto       3   
2       Uttar Pradesh   Central       Automobile             Auto       3   
3           Karnataka  Southern     Construction             Auto       2   
4             Gujarat   Western  Food Processing             Auto       2   
...               ...       ...              ...              ...     ...   
11246     Maharashtra   Western         Chemical           Office       4   
11247         Haryana  Northern       Healthcare       Veterinary       3   
11248  Madhya Pradesh   Central          Textile           Office       4   
11249       Karnataka  Southern      Agriculture           Office       3   
11250     Maharashtra   Western       Healthcare           Office       3   

        Amount  Status  unnamed1  
0      23952.0     NaN       NaN  
1      23934.0     NaN       NaN  
2      23924.0     NaN       NaN  
3      23912.0     NaN       NaN  
4      23877.0     NaN       NaN  
...        ...     ...       ...  
11246    370.0     NaN       NaN  
11247    367.0     NaN       NaN  
11248    213.0     NaN       NaN  
11249    206.0     NaN       NaN  
11250    188.0     NaN       NaN  

[11251 rows x 15 columns]

In [4]:
sales.head()

User_ID  Cust_name Product_ID Gender Age Group  Age  Marital_Status  \
0  1002903  Sanskriti  P00125942      F     26-35   28               0   
1  1000732     Kartik  P00110942      F     26-35   35               1   
2  1001990      Bindu  P00118542      F     26-35   35               1   
3  1001425     Sudevi  P00237842      M      0-17   16               0   
4  1000588       Joni  P00057942      M     26-35   28               1   

            State      Zone       Occupation Product_Category  Orders  \
0     Maharashtra   Western       Healthcare             Auto       1   
1  Andhra Pradesh  Southern             Govt             Auto       3   
2   Uttar Pradesh   Central       Automobile             Auto       3   
3       Karnataka  Southern     Construction             Auto       2   
4         Gujarat   Western  Food Processing             Auto       2   

    Amount  Status  unnamed1  
0  23952.0     NaN       NaN  
1  23934.0     NaN       NaN  
2  23924.0     NaN       NaN  
3  23912.0     NaN       NaN  
4  23877.0     NaN       NaN

Summary of DataFrame


In [5]:
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11251 entries, 0 to 11250
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   User_ID           11251 non-null  int64  
 1   Cust_name         11251 non-null  object 
 2   Product_ID        11251 non-null  object 
 3   Gender            11251 non-null  object 
 4   Age Group         11251 non-null  object 
 5   Age               11251 non-null  int64  
 6   Marital_Status    11251 non-null  int64  
 7   State             11251 non-null  object 
 8   Zone              11251 non-null  object 
 9   Occupation        11251 non-null  object 
 10  Product_Category  11251 non-null  object 
 11  Orders            11251 non-null  int64  
 12  Amount            11239 non-null  float64
 13  Status            0 non-null      float64
 14  unnamed1          0 non-null      float64
dtypes: float64(3), int64(4), object(8)
memory usage: 1.3+ MB


Statistical summary of the sales 

In [3]:
sales.describe()

User_ID           Age  Marital_Status        Orders        Amount  \
count  1.125100e+04  11251.000000    11251.000000  11251.000000  11239.000000   
mean   1.003004e+06     35.421207        0.420318      2.489290   9453.610858   
std    1.716125e+03     12.754122        0.493632      1.115047   5222.355869   
min    1.000001e+06     12.000000        0.000000      1.000000    188.000000   
25%    1.001492e+06     27.000000        0.000000      1.500000   5443.000000   
50%    1.003065e+06     33.000000        0.000000      2.000000   8109.000000   
75%    1.004430e+06     43.000000        1.000000      3.000000  12675.000000   
max    1.006040e+06     92.000000        1.000000      4.000000  23952.000000   

       Status  unnamed1  
count     0.0       0.0  
mean      NaN       NaN  
std       NaN       NaN  
min       NaN       NaN  
25%       NaN       NaN  
50%       NaN       NaN  
75%       NaN       NaN  
max       NaN       NaN

In [6]:
sales.dtypes

User_ID               int64
Cust_name            object
Product_ID           object
Gender               object
Age Group            object
Age                   int64
Marital_Status        int64
State                object
Zone                 object
Occupation           object
Product_Category     object
Orders                int64
Amount              float64
Status              float64
unnamed1            float64
dtype: object

Checking for null values:

In [7]:
sales.isnull().sum()

User_ID                 0
Cust_name               0
Product_ID              0
Gender                  0
Age Group               0
Age                     0
Marital_Status          0
State                   0
Zone                    0
Occupation              0
Product_Category        0
Orders                  0
Amount                 12
Status              11251
unnamed1            11251
dtype: int64

Since I found out there are null values, I'll have to clean the data set
-  Cleaning The Data Set:

Status and Unnamed1 columns have null values so I have to drop them:

In [12]:
sales_cleaned=sales.drop(columns=['Status','unnamed1'])
sales_cleaned

User_ID    Cust_name Product_ID Gender Age Group  Age  Marital_Status  \
0      1002903    Sanskriti  P00125942      F     26-35   28               0   
1      1000732       Kartik  P00110942      F     26-35   35               1   
2      1001990        Bindu  P00118542      F     26-35   35               1   
3      1001425       Sudevi  P00237842      M      0-17   16               0   
4      1000588         Joni  P00057942      M     26-35   28               1   
...        ...          ...        ...    ...       ...  ...             ...   
11246  1000695      Manning  P00296942      M     18-25   19               1   
11247  1004089  Reichenbach  P00171342      M     26-35   33               0   
11248  1001209        Oshin  P00201342      F     36-45   40               0   
11249  1004023       Noonan  P00059442      M     36-45   37               0   
11250  1002744      Brumley  P00281742      F     18-25   19               0   

                State      Zone       Occupation Product_Category  Orders  \
0         Maharashtra   Western       Healthcare             Auto       1   
1      Andhra Pradesh  Southern             Govt             Auto       3   
2       Uttar Pradesh   Central       Automobile             Auto       3   
3           Karnataka  Southern     Construction             Auto       2   
4             Gujarat   Western  Food Processing             Auto       2   
...               ...       ...              ...              ...     ...   
11246     Maharashtra   Western         Chemical           Office       4   
11247         Haryana  Northern       Healthcare       Veterinary       3   
11248  Madhya Pradesh   Central          Textile           Office       4   
11249       Karnataka  Southern      Agriculture           Office       3   
11250     Maharashtra   Western       Healthcare           Office       3   

        Amount  
0      23952.0  
1      23934.0  
2      23924.0  
3      23912.0  
4      23877.0  
...        ...  
11246    370.0  
11247    367.0  
11248    213.0  
11249    206.0  
11250    188.0  

[11251 rows x 13 columns]

The column 'Amount' has a few null values so I'll replace them with the median of the column
- First I calculate the median of the 'Amount' column
- Then replace the null values with the median


In [13]:
amount_median=sales_cleaned['Amount'].median()

sales_cleaned['Amount'].fillna(amount_median, inplace=True)

C:\Users\use\AppData\Local\Temp\ipykernel_1932\2628319314.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  sales_cleaned['Amount'].fillna(amount_median, inplace=True)


In [20]:
print(amount_median)

8109.0


In [14]:
sales_cleaned.isnull().sum()

User_ID             0
Cust_name           0
Product_ID          0
Gender              0
Age Group           0
Age                 0
Marital_Status      0
State               0
Zone                0
Occupation          0
Product_Category    0
Orders              0
Amount              0
dtype: int64

## Visualizations

In [16]:
pip install plotly

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [17]:
import pandas as pd
import numpy as np
import plotly.express as px


In [18]:
sales_cleaned.columns

Index(['User_ID', 'Cust_name', 'Product_ID', 'Gender', 'Age Group', 'Age',
       'Marital_Status', 'State', 'Zone', 'Occupation', 'Product_Category',
       'Orders', 'Amount'],
      dtype='object')

# Heat Map

- Selecting the columns with numerals to craete a heatmap for correlation calculation 
- Creating the heatmap

In [19]:
numerals_data = sales_cleaned.select_dtypes(include=[np.number])
correlation_matrix = numerals_data.corr()

fig = px.imshow(correlation_matrix,
                text_auto=True,  
                labels=dict(x="Features", y="Features", color="Correlation"),
                x=correlation_matrix.columns,
                y=correlation_matrix.columns)
fig.update_layout(title="Correlation Matrix Heatmap",
                  xaxis_title="Features",
                  yaxis_title="Features")

fig.show()

From the heatmap we can conclude that:
- Age and User_ID have a weak positive correlation which means that as age increases the User_ID increases in a small amount.ETC 

In [26]:
sales_cleaned.head()

User_ID  Cust_name Product_ID Gender Age Group  Age  Marital_Status  \
0  1002903  Sanskriti  P00125942      F     26-35   28               0   
1  1000732     Kartik  P00110942      F     26-35   35               1   
2  1001990      Bindu  P00118542      F     26-35   35               1   
3  1001425     Sudevi  P00237842      M      0-17   16               0   
4  1000588       Joni  P00057942      M     26-35   28               1   

            State      Zone       Occupation Product_Category  Orders   Amount  
0     Maharashtra   Western       Healthcare             Auto       1  23952.0  
1  Andhra Pradesh  Southern             Govt             Auto       3  23934.0  
2   Uttar Pradesh   Central       Automobile             Auto       3  23924.0  
3       Karnataka  Southern     Construction             Auto       2  23912.0  
4         Gujarat   Western  Food Processing             Auto       2  23877.0

## Bar Plot:Age Group vs Orders

- Creating a Bar plot of age groups vs orders so as to know the purchasing behavior among age groups

In [49]:

fig = px.bar(sales_cleaned, x='Age Group',y='Orders',title='Bar plot for purchasing power',color_discrete_sequence=['blue'])

fig.update_layout(
    plot_bgcolor='lightgray', 
    paper_bgcolor='lightgray')  

fig.show()


From the bar plot we can conclude:
- The highest purchasing power is between the age group of 26-35
- The lowest purchasing power is between the age group of 0-17


## Pie Chart 

Checking states with the highest sales volume during the Diwali season

In [37]:
fig = px.pie(sales_cleaned,values='Amount',names='State',title='States with the highest volume sales')

fig.show()

From the pie chart we can conclude that:
- The state of Uttar Pradesh has the highest number of volume sales
- The state of Telangana has the least number of volume sales

## Histogram

Creating a histogram to know the top selling product categories

In [39]:
fig = px.histogram(sales_cleaned, x='Product_Category', color='Orders' ,nbins=5, title='Top selling product categories')

fig.show()


From the histogram we can conclude:
- Clothing and Apparel category is the top selling category with 659 people making 4 orders
- Hand and power tools category is the least selling category with only 12 people making 4 orders

## Box Plot

Creating a box plot to show how occupation influences spending patterns

In [40]:
fig = px.box(sales_cleaned, x='Occupation', y='Amount', title='Influence of Occupation in Spending patterns')

fig.show()

From the box plot we can find out that:
- The government employees tend to spend more money since the maximum amount which has been used by them is above 23k almost reaching 24k 
- The people who have an occupation under Healthcare have the least amount spent which is a minimum of 188


## Scatter Plot

Creating a scatter plot showing spending between genders and how marital status affects it

In [47]:
fig = px.scatter(sales_cleaned, x='Gender', y='Amount',color='Marital_Status', title='Influence of gender on spending with marital status')

fig.update_layout(
    width=800,  
    height=800   
)

fig.show()

From the scatter plot we can conclude that:
- The gender which has spent the most money is surprisingly the female gender for which they are not married (beating males by a very small mergin)
- Females also record the least amount of money spent spending as low as 188 and of course they are not married
